File: 0111-Minimum-Depth-of-Binary-Tree/0111.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public int minDepth(TreeNode root) {
        return minHeight(root);
    }

    public int minHeight(TreeNode node) {
        if (node == null)
            return 0;

        int left = minHeight(node.left);
        int right = minHeight(node.right);

        if (left > 0 && right > 0) {
            return Math.min(left, right) + 1;
        } else if (left > 0) {
            return left + 1;
        } else {
            return right + 1;
        }
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

File: 0111-Minimum-Depth-of-Binary-Tree/0111.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int: # type: ignore
        
        def min_height(root):
            if not root: return 0

            left = min_height(root.left)
            right = min_height(root.right)

            if left > 0 and right > 0:
                return 1 + min(left, right)
            elif left > 0:
                return 1 + left
            else:
                return 1 + right

        return min_height(root)